crates/ra_syntax/src/ast/traits.rs

```diff
@@ -69,16 +75,47 @@ pub trait ModuleItemOwner: AstNode {
 #[derive(Debug)]
 pub struct ItemOrMacroIter(SyntaxNodeChildren);
 
+fn collect_derive_attr(n: impl AttrsOwner) -> Option<ItemOrMacro> {
+    let attrs = n.attrs();
+
+    for attr in attrs {
+        if let Some((name, _)) = attr.as_call() {
+            if name.as_str() == "derive" {
```

Comment:
this relates to the semantics of a particular attribute. I think we shouldn't chage anything about ra_syntax at this point: all atrributes are already in the syntax tree, they sit on top of the items.

Rather, in hir::nameres::raw, when we process an item, we should convert its attributes to token trees and attach the token trees to the correspondin raw item. Then, in the nameres::collector, we an actually look at thouse attrs, and expand them as macros